File: python_implementations/solutions/l2.py
import threading


def multiply_split_row(a, b, result, start, end):
    rows_a, cols_a = len(a), len(a[0])
    rows_b, cols_b = len(b), len(b[0])
    for i in range(start, end):
        for j in range(cols_b):
            result[i][j] = sum(a[i][k] * b[k][j] for k in range(len(b)))


def multiply_multithreading_row(a, b, num_threads):
    rows_a, cols_a = len(a), len(a[0])
    rows_b, cols_b = len(b), len(b[0])

    if cols_a != rows_b:
        return "The matrices are not compatible for multiplication."

    result = [[0 for _ in range(cols_b)] for _ in range(rows_a)]
    threads = []

    rows_per_thread = rows_a // num_threads
    for i in range(num_threads):
        start = i * rows_per_thread
        # The last thread should get the remaining rows if the number of rows is not divisible by num_threads
        end = rows_a if i == num_threads - 1 else start + rows_per_thread
        thread = threading.Thread(target=multiply_split_row, args=(a, b, result, start, end))
        threads.append(thread)
        thread.start()

    for thread in threads:
        thread.join()

    return result


import math


def multiply_block(a, b, result, start_row, end_row, start_col, end_col):
    for i in range(start_row, end_row):
        for j in range(start_col, end_col):
            result[i][j] = sum(a[i][k] * b[k][j] for k in range(len(b)))


def multiply_multithreading_block(a, b, num_threads):
    rows_a, cols_a = len(a), len(a[0])
    rows_b, cols_b = len(b), len(b[0])

    if cols_a != rows_b:
        return "The matrices are not compatible for multiplication."

    result = [[0 for _ in range(cols_b)] for _ in range(rows_a)]
    threads = []

    block_size = int(math.sqrt(num_threads))
    rows_per_block = rows_a // block_size
    cols_per_block = cols_b // block_size

    for i in range(block_size):
        for j in range(block_size):
            start_row = i * rows_per_block
            end_row = rows_a if i == block_size - 1 else start_row + rows_per_block
            start_col = j * cols_per_block
            end_col = cols_b if j == block_size - 1 else start_col + cols_per_block
            thread = threading.Thread(target=multiply_block,
                                      args=(a, b, result, start_row, end_row, start_col, end_col))
            threads.append(thread)
            thread.start()

    for thread in threads:
        thread.join()

    return result

File: cpp_implementations/solutions/l2.cpp
#include <vector>
#include <thread>
#include <cmath>

using namespace std;

void multiply_rows(const vector<vector<int>>& a, const vector<vector<int>>& b, vector<vector<int>>& result, int start, int end) {
    for (int i = start; i < end; ++i) {
        for (size_t j = 0; j < b[0].size(); ++j) {
            for (size_t k = 0; k < b.size(); ++k) {
                result[i][j] += a[i][k] * b[k][j];
            }
        }
    }
}

vector<vector<int>> multiplyMultithreadingRows(const vector<vector<int>>& a, const vector<vector<int>>& b, int num_threads) {
    vector<vector<int>> result(a.size(), vector<int>(b[0].size(), 0));
    vector<thread> threads;

    int rows_per_thread = a.size() / num_threads;
    for (int i = 0; i < num_threads; ++i) {
        int start = i * rows_per_thread;
        int end = (i == num_threads - 1) ? a.size() : start + rows_per_thread;
        threads.emplace_back(multiply_rows, cref(a), cref(b), ref(result), start, end);
    }

    for (auto& thread : threads) {
        thread.join();
    }

    return result;
}


void multiply_block(const vector<vector<int>>& a, const vector<vector<int>>& b, vector<vector<int>>& result, int start_row, int end_row, int start_col, int end_col) {
    for (int i = start_row; i < end_row; ++i) {
        for (int j = start_col; j < end_col; ++j) {
            for (size_t k = 0; k < b.size(); ++k) {
                result[i][j] += a[i][k] * b[k][j];
            }
        }
    }
}

vector<vector<int>> multiplyMultithreadingBlocks(const vector<vector<int>>& a, const vector<vector<int>>& b, int num_threads) {
    vector<vector<int>> result(a.size(), vector<int>(b[0].size(), 0));
    vector<thread> threads;

    int block_size = sqrt(num_threads);
    int rows_per_block = a.size() / block_size;
    int cols_per_block = b[0].size() / block_size;

    for (int i = 0; i < block_size; ++i) {
        for (int j = 0; j < block_size; ++j) {
            int start_row = i * rows_per_block;
            int end_row = (i == block_size - 1) ? a.size() : start_row + rows_per_block;
            int start_col = j * cols_per_block;
            int end_col = (j == block_size - 1) ? b[0].size() : start_col + cols_per_block;
            threads.emplace_back(multiply_block, cref(a), cref(b), ref(result), start_row, end_row, start_col, end_col);
        }
    }

    for (auto& thread : threads) {
        thread.join();
    }

    return result;
}